crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -1084,12 +996,212 @@ impl<'a, 'o> TimelineEventHandler<'a, 'o> {
+        // non-decreasing order.
+
+        #[derive(Debug)]
+        enum Operation {
+            Insert(usize, MilliSecondsSinceUnixEpoch),
+            Replace(usize, MilliSecondsSinceUnixEpoch),
+            Remove(usize),
+        }
+
+        let mut ops = vec![];
+
+        let mut prev_item: Option<(&Arc<TimelineItem>, MilliSecondsSinceUnixEpoch)> = None;
+        let mut last_event_ts = None;
+
+        for (i, item) in self.items.iter().enumerate() {
+            if let TimelineItemKind::Virtual(VirtualTimelineItem::DayDivider(ts)) = item.kind() {
+                // It's a day divider.
+
+                if let Some((prev_item, prev_ts)) = prev_item {
+                    if prev_item.is_day_divider() {
+                        trace!("removing duplicate day divider @ {i}");
+                        // This day divider is preceded by another one: remove the current one.
+                        ops.push(Operation::Remove(i));
+                    } else if prev_item.is_event() {
+                        // This day divider is preceded by an event.
+                        if timestamp_to_date(prev_ts) == timestamp_to_date(*ts) {
+                            // The event has the same date as the day divider: remove the day
+                            // divider.
+                            trace!(
+                                "removing day divider following event with same timestamp @ {i}"
+                            );
+                            ops.push(Operation::Remove(i));
+                        }
+                    }
+                } else {
+                    // No interesting item prior to the day divider: it must be
+                    // the first one, nothing to do.
+                }
+
+                prev_item = Some((item, *ts));
+            }
+
+            if let Some(event) = item.as_event() {
+                // It's an event.
+                let ts = event.timestamp();
+
+                if let Some((prev_item, prev_ts)) = prev_item {
+                    if prev_item.is_day_divider() {
+                        // The event is preceded by a day divider.
+                        if timestamp_to_date(prev_ts) != timestamp_to_date(ts) {
+                            // The day divider is wrong. Should we replace it with the correct
+                            // value, or remove it entirely?
+                            //
+                            // Look at the previous event timestamp: if they became the same value,
+                            // remove the divider entirely. Otherwise, they show different values,
+                            // so keep the day divider but adjust its value.
+                            if last_event_ts.map_or(
+                                false,
+                                |last_event_ts: MilliSecondsSinceUnixEpoch| {
+                                    timestamp_to_date(last_event_ts) == timestamp_to_date(ts)
+                                },
```

Comment:
If only we had `goto` :smiling_face_with_tear: 

In my initial version, I had this:

```rust
                                let mut removed = false;

                                if let Some(last_event_ts) = latest_event_ts {
                                    if timestamp_to_date(last_event_ts) == event_date {
                                        // There's a previous event with the same date: remove the divider.
                                        trace!("removed day divider @ {i} between two events that have the same date");
                                        ops.push(DayDividerOperation::Remove(i - 1));
                                        removed = true;
                                    }
                                }

                                if !removed {
                                    // There's no previous event or there's one with a different date:
                                    // replace the current divider.
                                    trace!(
                                        "replacing day divider @ {i} with new timestamp from event"
                                    );
                                    ops.push(DayDividerOperation::Replace(i - 1, ts));
                                }
```

I slightly prefer the `map_or` for its conciseness, but since you asked for my initial version, I'll happily revert to it :)